source/solvers/navier_stokes_base.cc

```diff
@@ -1789,79 +1820,168 @@ NavierStokesBase<dim, VectorType, DofsType>::establish_solid_domain(
+                      stasis_constraint_struct
+                        .filtered_phase_fraction_tolerance)
+                    {
+                      cell_is_in_right_fluid = false;
+                      break;
+                    }
                 }
+
+              // If the cell is not in the right fluid, no solid constraint will
+              // be applied on the cell's DOFs. We flag the current cell's DOFs
+              // as "connected to fluid" and we skip to the next cell.
+              if (!cell_is_in_right_fluid)
+                {
+                  flag_dofs_connected_to_fluid(
+                    local_dof_indices,
+                    stasis_constraint_struct.dofs_are_connected_to_fluid);
+                  continue;
+                }
+
+              get_cell_temperature_values(cell,
+                                          dof_handler_ht,
+                                          temperature_solution,
+                                          local_temperature_values);
+              add_flags_and_constrain_velocity(local_dof_indices,
+                                               local_temperature_values,
+                                               stasis_constraint_struct);
             }
-          // Flag non-solid cell DOFs to identify which pressure DOFs should not
-          // be constrained.
-          if (!cell_is_solid)
-            flag_dofs_connected_to_fluid(local_dof_indices,
-                                         dofs_are_connected_to_fluid);
         }
+      check_and_constrain_pressure(stasis_constraint_struct, local_dof_indices);
     }
+}
 
+template <int dim, typename VectorType, typename DofsType>
+void
+NavierStokesBase<dim, VectorType, DofsType>::add_flags_and_constrain_velocity(
+  const std::vector<types::global_dof_index> &local_dof_indices,
+  const std::vector<double>                  &local_temperature_values,
+  StasisConstraintWithTemperature            &stasis_constraint_struct)
+{
+  bool cell_is_solid = false;
+
+  // Flag cell if the cell is solid and constrain its velocity DOFs
+  for (const double &temperature : local_temperature_values)
+    {
+      if (temperature >= stasis_constraint_struct.min_solid_temperature &&
+          temperature <= stasis_constraint_struct.max_solid_temperature)
+        {
+          flag_dofs_in_solid(local_dof_indices,
+                             stasis_constraint_struct.dofs_are_in_solid);
+          constrain_solid_cell_velocity_dofs(false,
+                                             local_dof_indices,
+                                             this->dynamic_zero_constraints);
+          cell_is_solid = true;
+          break;
+        }
+    }
+
+  // Flag non-solid cell DOFs to identify which pressure DOFs should not
+  // be constrained.
+  if (!cell_is_solid)
+    flag_dofs_connected_to_fluid(
+      local_dof_indices, stasis_constraint_struct.dofs_are_connected_to_fluid);
```

Comment:
Check here if the if condition for solid flag is ok. In my head, if 1 quadrature point is solid than the cell should not be flag solid.